### ISSUE TITLE
Edge case: complex torch.nn.PoissonNLLLoss has inconsistent NaN behavior between CPU and CUDA

### ISSUE THREAD

**Issue by triumph-wangyuyang** [pytorch/pytorch#89403]
### 🐛 Describe the bug

Test on the CPU:

    import torch
    loss = torch.nn.PoissonNLLLoss()
    log_input = torch.rand([5, 0], dtype=torch.complex64)
    target = torch.rand([5, 1], dtype=torch.complex64)
    out = loss(log_input, target)
    print(out)

result: tensor(nan+nanj)

Test on the GPU:

    import torch
    loss = torch.nn.PoissonNLLLoss().cuda()
    log_input = torch.rand([5, 0], dtype=torch.complex64).cuda()
    target = torch.rand([5, 1], dtype=torch.complex64).cuda()
    out = loss(log_input, target)
    print(out)

result: tensor(nan+0.j , device='cuda:0')

Why is nanj running on the CPU, but 0j running on the GPU?

### Versions

pytorch: 1.12.1
Python version: 3.8
CUDA/cuDNN version: cuDNN 11.3
GPU models and configuration: RTX3060
Operating System：Windows

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved

**Comment by lezcano** [comment]
This still reproduces in master. That being said, note that it's not that it returns `0j` but `nan+0j`, which is still a `nan` value. This looks fine to me. What do you think this function should return?

**Comment by anjali411** [comment]
Hmm I don't know why the imaginary value on CUDA is not nan but both values represent nan:
```
>>> cmath.isnan(float('nan')+0j)
True
>>> cmath.isnan(float('nan') * (1+1j))
True
```

**Comment by lezcano** [comment]
This seems fair to me. @triumph-wangyuyang what use case do you have that having these two representaitons of nan is problematic?

**Comment by triumph-wangyuyang** [comment]
> This seems fair to me. @triumph-wangyuyang what use case do you have that having these two representaitons of nan is problematic?

Previously in testing, I found that having two nans resulted in incorrect calculation precision.

**Comment by lezcano** [comment]
Could you provide a snippet reproducng the issue? Perhaps using a different treatment for nans in your code we could make it robust enough so that this discrepancy is not an issue.

**Comment by triumph-wangyuyang** [comment]
> Could you provide a snippet reproducng the issue? Perhaps using a different treatment for nans in your code we could make it robust enough so that this discrepancy is not an issue.

Sorry, I can't restore and reproduce this problem for the time being, I didn't save the source file after I found the problem.

**Comment by lezcano** [comment]
I'm closing it for now, as both functions seem to return nans, even if the nans are a bit different. Feel free to reopen if this is still an issue for your application
